### ISSUE TITLE
Reflector key conflict across namespaces

### ISSUE THREAD

**Issue by gnieto** [clux/kube-rs#55]
Hi! 

I was using this library and I found that `Reflector` indexes the objects by `name`. If a reflector is used without `within` method, as keys are indexed by `name`, if a `name` is shared among distinct `namespace`s, some objects gets overwritten on the `Cache`.

For example, with the given code:

```rust
    let config = kube_config::load_kube_config().expect("failed to load kubeconfig");
    let client = APIClient::new(config);
    let resource = Api::v1Pod(client);
    let rf = Reflector::new(resource).init().unwrap();

    let cache = rf.read().unwrap();
    for c in cache.iter() {
        println!("KEY: {} {:?}", c.0, c.1.metadata.namespace);
    } 
```

I get the following results:

```
KEY: hello-node Some("default")
KEY: hello-node-55b49fb9f8-dcfdp Some("default")
KEY: hello-node-never-restart Some("test")
```
With  `kubectl get pod --all-namespaces` I get:

```
default       hello-node                         0/1     CrashLoopBackOff   236        10d
default       hello-node-55b49fb9f8-dcfdp        1/1     Running            0          11d
default       hello-node-never-restart           1/1     Running            0          10d
test          hello-node-never-restart           1/1     Running            0          33m
```

I was expecting this cache to be indexed by both `name` and `namespace`, which I think is a unique identifier for objects (but I'm not sure about this).

**Comment by clux** [comment]
Damnit. Yeah, this is a legit bug. Thank you.

The cache is currently a pretty dumb Map of `name -> Object` atm. Remember thinking that "maybe I should keep this as a vector", but then concluded "nahhh, `metadata.name` is unique within a namespace, right"... :disappointed: 

#### Potential fixes

- convert the cache to a `Vec<K>` and __maybe later__ add a .get() method on the reflector
- extend the key of the Map to be a pair: `(name, namespace) -> Object`
- ~~prefix/suffix namespace to key~~ (no)

Leaning heavily towards the first here, because the map options all seem to have drawbacks: they'll either require some ad-hoc magic to disassemble the parts, or awkward double pattern matching in for loops to get the name + object. May as well just inspect what is there anyway.

**Comment by gnieto** [comment]
Using a `Vec` means that all the lookups to add/update/remove once a `WatchEvent` is handled requires to read all the `Vec` and it will be harder to work with.

Would do you think about maintaining the `Cache` with a `BTreeMap` internal, indexed by `(name, namespace)` and add a method to get an iterator over the objects instead? 

If you are ok with this, I can prepare a PR.

**Comment by clux** [comment]
Yeah true. It would be easier to implement there like that, and then people can easily just iterate over `.values()` externally if that's more convenient than the double pattern match.

You could technically hide that crud complexity within `Reflector` since it does not expose `WatchEvent` externally, but thinking more about it, there's not a great need to.

Yeah, go for it. If you can get it to compile with all the examples I can quickly make a release tomorrow!

**Comment by clux** [comment]
Released your pr in `kube@0.14.0`. Thanks a lot!
